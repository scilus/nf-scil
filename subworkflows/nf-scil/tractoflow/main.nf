// PREPROCESSING
include { ANATOMICAL_SEGMENTATION                                  } from '../anatomical_segmentation/main'
include { PREPROC_DWI                                              } from '../preproc_dwi/main'
include { PREPROC_T1                                               } from '../preproc_t1/main'
include { REGISTRATION as T1_REGISTRATION                          } from '../registration/main'
include { RECONST_DTIMETRICS as REGISTRATION_FA                    } from '../../../modules/nf-scil/reconst/dtimetrics/main'
include { REGISTRATION_ANTSAPPLYTRANSFORMS as TRANSFORM_WMPARC     } from '../../../modules/nf-scil/registration/antsapplytransforms/main'
include { REGISTRATION_ANTSAPPLYTRANSFORMS as TRANSFORM_APARC_ASEG } from '../../../modules/nf-scil/registration/antsapplytransforms/main'

// RECONSTRUCTION
include { RECONST_FRF        } from '../../../modules/nf-scil/reconst/frf/main'
include { RECONST_MEANFRF    } from '../../../modules/nf-scil/reconst/meanfrf/main'
include { RECONST_DTIMETRICS } from '../../../modules/nf-scil/reconst/dtimetrics/main'
include { RECONST_FODF       } from '../../../modules/nf-scil/reconst/fodf/main'

// TRACKING
include { TRACKING_PFTTRACKING   } from '../../../modules/nf-scil/tracking/pfttracking/main'
include { TRACKING_LOCALTRACKING } from '../../../modules/nf-scil/tracking/localtracking/main'

params.dwi_fodf_fit_use_average_frf = false

workflow TRACTOFLOW {

    take:
        ch_dwi // channel: [ val(meta), dwi_nifti, bval_file, bvec_file ]
        ch_sbref // channel: [ val(meta), sbref_nifti ], optional
        ch_rev_dwi // channel: [ val(meta), rev_dwi_nifti, bval_file, bvec_file ], optional
        ch_rev_sbref // channel: [ val(meta), rev_sbref_nifti ], optional
        ch_t1 // channel: [ val(meta), t1_nifti ]
        ch_wmparc // channel: [ val(meta), wmparc_nifti ], optional
        ch_aparc_aseg // channel: [ val(meta), aparc+aseg_nifti ], optional
        ch_topup_config // channel: [ topup_config_file ], optional
        ch_bet_template // channel: [ bet_template_nifti ]
        ch_bet_probability // channel: [ bet_template_nifti ]

    main:

    ch_versions = Channel.empty()

    /* Unpack inputs */
    ch_inputs = ch_dwi
        .join(ch_sbref, remainder: true)
        .join(ch_rev_dwi, remainder: true)
        .map{ it[5] ? it : it[0..4] + [null, null, null] }
        .join(ch_rev_sbref, remainder: true)
        .join(ch_t1, remainder: true)
        .join(ch_wmparc, remainder: true)
        .join(ch_aparc_aseg, remainder: true)
        .view()
        .multiMap{meta, dwi, bval, bvec, sbref, rev_dwi, rev_bval, rev_bvec, rev_sbref, t1, wmparc, aparc_aseg ->
            dwi: [meta, dwi, bval, bvec]
            sbref: sbref ? [meta, sbref] : []
            rev_dwi: rev_dwi ? [meta, rev_dwi, rev_bval, rev_bvec] : []
            rev_sbref: rev_sbref ? [meta, rev_sbref] : []
            t1: [meta, t1]
            wmparc: wmparc ? [meta, wmparc] : []
            aparc_aseg: aparc_aseg ? [meta, aparc_aseg] : []
        }

    /* PREPROCESSING */

    //
    // SUBWORKFLOW: Run PREPROC_DWI
    //
    PREPROC_DWI(
        ch_inputs.dwi, ch_inputs.rev_dwi.filter{ it },
        ch_inputs.sbref.filter{ it }, ch_inputs.rev_sbref.filter{ it },
        ch_topup_config.ifEmpty( "b02b0.cnf" )
    )
    ch_versions = ch_versions.mix(PREPROC_DWI.out.versions.first())

    //
    // SUBWORKFLOW: Run PREPROC_T1
    //
    ch_t1_meta = ch_inputs.t1.map{ it[0] }
    PREPROC_T1(
        ch_inputs.t1,
        ch_t1_meta.combine(ch_bet_template),
        ch_t1_meta.combine(ch_bet_probability),
        Channel.empty(),
        Channel.empty(),
        Channel.empty()
    )
    ch_versions = ch_versions.mix(PREPROC_T1.out.versions.first())

    //
    // MODULE: Run RECONST/DTIMETRICS (REGISTRATION_FA)
    //
    ch_registration_fa = PREPROC_DWI.out.dwi_resample
        .join(PREPROC_DWI.out.bval)
        .join(PREPROC_DWI.out.bvec)
        .join(PREPROC_DWI.out.b0_mask)
    REGISTRATION_FA( ch_registration_fa )
    ch_versions = ch_versions.mix(REGISTRATION_FA.out.versions.first())

    //
    // SUBWORKFLOW: Run REGISTRATION
    //
    T1_REGISTRATION(
        PREPROC_T1.out.t1_final,
        PREPROC_DWI.out.b0,
        REGISTRATION_FA.out.fa,
        []
    )
    ch_versions = ch_versions.mix(T1_REGISTRATION.out.versions.first())

    /* SEGMENTATION */

    //
    // MODULE: Run REGISTRATION_ANTSAPPLYTRANSFORMS (TRANSFORM_WMPARC)
    //
    TRANSFORM_WMPARC(
        ch_inputs.wmparc
            .filter{ it[1] }
            .join(PREPROC_DWI.out.b0)
            .join(T1_REGISTRATION.out.transfo_image)
            .map{ it[0..2] + [it[3..-1]] }
    )
    ch_versions = ch_versions.mix(TRANSFORM_WMPARC.out.versions.first())

    //
    // MODULE: Run REGISTRATION_ANTSAPPLYTRANSFORMS (TRANSFORM_APARC_ASEG)
    //
    TRANSFORM_APARC_ASEG(
        ch_inputs.aparc_aseg
            .filter{ it[1] }
            .join(PREPROC_DWI.out.b0)
            .join(T1_REGISTRATION.out.transfo_image)
            .map{ it[0..2] + [it[3..-1]] }
    )
    ch_versions = ch_versions.mix(TRANSFORM_APARC_ASEG.out.versions.first())

    //
    // SUBWORKFLOW: Run ANATOMICAL_SEGMENTATION
    //
    ANATOMICAL_SEGMENTATION(
        T1_REGISTRATION.out.image_warped,
        TRANSFORM_WMPARC.out.warped_image
            .join(TRANSFORM_APARC_ASEG.out.warped_image),
        Channel.empty()
    )
    ch_versions = ch_versions.mix(ANATOMICAL_SEGMENTATION.out.versions.first())

    /* RECONSTRUCTION */

    //
    // MODULE: Run RECONST/DTIMETRICS
    //
    ch_dti_metrics = PREPROC_DWI.out.dwi_resample
        .join(PREPROC_DWI.out.bval)
        .join(PREPROC_DWI.out.bvec)
        .join(PREPROC_DWI.out.b0_mask)
    RECONST_DTIMETRICS( ch_dti_metrics )
    ch_versions = ch_versions.mix(RECONST_DTIMETRICS.out.versions.first())

    //
    // MODULE: Run RECONST/FRF
    //
    ch_reconst_frf = PREPROC_DWI.out.dwi_resample
        .join(PREPROC_DWI.out.bval)
        .join(PREPROC_DWI.out.bvec)
        .join(PREPROC_DWI.out.b0_mask)
        .join(ANATOMICAL_SEGMENTATION.out.wm_mask)
        .join(ANATOMICAL_SEGMENTATION.out.gm_mask)
        .join(ANATOMICAL_SEGMENTATION.out.csf_mask)
    RECONST_FRF( ch_reconst_frf )
    ch_versions = ch_versions.mix(RECONST_FRF.out.versions.first())

    /* Run fiber response aeraging over subjects */
    ch_fiber_response = RECONST_FRF.out.frf.map{ it.flatten() }
    if ( params.dwi_fodf_fit_use_average_frf ) {
        RECONST_MEANFRF( RECONST_FRF.out.frf.map{ it[1] }.flatten() )
        ch_versions = ch_versions.mix(RECONST_MEANFRF.out.versions.first())
        ch_fiber_response = RECONST_FRF.out.map{ it[0] }
            .combine( RECONST_MEANFRF.out.meanfrf )
    }

    ch_fiber_response = ch_fiber_response
        .map{ it.size() > 2 ? it : it + [[], []] }

    //
    // MODULE: Run RECONST/FODF
    //
    ch_reconst_fodf = PREPROC_DWI.out.dwi_resample
        .join(PREPROC_DWI.out.bval)
        .join(PREPROC_DWI.out.bvec)
        .join(PREPROC_DWI.out.b0_mask)
        .join(RECONST_DTIMETRICS.out.fa)
        .join(RECONST_DTIMETRICS.out.md)
        .join(ch_fiber_response)
    RECONST_FODF( ch_reconst_fodf )
    ch_versions = ch_versions.mix(RECONST_FODF.out.versions.first())

    //
    // MODULE: Run TRACKING/PFTTRACKING
    //
    ch_pft_tracking = ANATOMICAL_SEGMENTATION.out.wm_mask
        .join(ANATOMICAL_SEGMENTATION.out.gm_mask)
        .join(ANATOMICAL_SEGMENTATION.out.csf_mask)
        .join(RECONST_FODF.out.fodf)
        .join(RECONST_DTIMETRICS.out.fa)
    TRACKING_PFTTRACKING( ch_pft_tracking )
    ch_versions = ch_versions.mix(TRACKING_PFTTRACKING.out.versions.first())

    //
    // MODULE: Run TRACKING/LOCALTRACKING
    //
    ch_local_tracking = ANATOMICAL_SEGMENTATION.out.wm_mask
        .join(RECONST_FODF.out.fodf)
        .join(RECONST_DTIMETRICS.out.fa)
    TRACKING_LOCALTRACKING( ch_local_tracking )
    ch_versions = ch_versions.mix(TRACKING_LOCALTRACKING.out.versions.first())

    // define output channels
    ch_dwi_preprocessed = PREPROC_DWI.out.dwi_resample
        .join(PREPROC_DWI.out.bval)
        .join(PREPROC_DWI.out.bvec)
    ch_brain_mask = PREPROC_DWI.out.b0_mask
    ch_t1_registered = T1_REGISTRATION.out.image_warped
    ch_t1_to_b0_transform = T1_REGISTRATION.out.transfo_image
        .join(T1_REGISTRATION.out.transfo_trk)
    ch_tissues_masks = ANATOMICAL_SEGMENTATION.out.wm_mask
        .join(ANATOMICAL_SEGMENTATION.out.gm_mask)
        .join(ANATOMICAL_SEGMENTATION.out.csf_mask)
    ch_tissues_maps = ANATOMICAL_SEGMENTATION.out.wm_map
        .join(ANATOMICAL_SEGMENTATION.out.gm_map)
        .join(ANATOMICAL_SEGMENTATION.out.csf_map)
    ch_dti_tensor = RECONST_DTIMETRICS.out.tensor
    ch_dti_metrics = RECONST_DTIMETRICS.out.ad
        .join(RECONST_DTIMETRICS.out.rd)
        .join(RECONST_DTIMETRICS.out.md)
        .join(RECONST_DTIMETRICS.out.fa)
        .join(RECONST_DTIMETRICS.out.ga)
        .join(RECONST_DTIMETRICS.out.rgb)
        .join(RECONST_DTIMETRICS.out.evecs_v1)
        .join(RECONST_DTIMETRICS.out.evecs_v2)
        .join(RECONST_DTIMETRICS.out.evecs_v3)
        .join(RECONST_DTIMETRICS.out.evals_e1)
        .join(RECONST_DTIMETRICS.out.evals_e2)
        .join(RECONST_DTIMETRICS.out.evals_e3)
        .join(RECONST_DTIMETRICS.out.residual)
        .join(RECONST_DTIMETRICS.out.nonphysical)
        .join(RECONST_DTIMETRICS.out.mode)
        .join(RECONST_DTIMETRICS.out.norm)
    ch_local_tractogram = TRACKING_LOCALTRACKING.out.trk
        .join(TRACKING_LOCALTRACKING.out.config)
    ch_pft_tractogram = TRACKING_PFTTRACKING.out.trk
        .join(TRACKING_PFTTRACKING.out.config)

    emit:
        dwi = ch_dwi_preprocessed
            // channel: [ val(meta), path(*.nii.gz), path(*.bval), path(*.bvec) ]
        brain_mask = ch_brain_mask
            // channel: [ val(meta), path(*.nii.gz) ]
        t1 = ch_t1_registered
            // channel: [ val(meta), path(*.nii.gz) ]
        t1_to_b0_transformations = ch_t1_to_b0_transform
            // channel: [ val(meta), [path(*.nii.gz | *.txt), ...], [path(*.nii.gz | *.txt), ...] ]
        tissues_masks = ch_tissues_masks
            // channel: [ val(meta), path(*.nii.gz), path(*.nii.gz), path(*.nii.gz) ]
        tissues_maps = ch_tissues_maps
            // channel: [ val(meta), path(*.nii.gz), path(*.nii.gz), path(*.nii.gz) ]
        dti_tensor = ch_dti_tensor
            // channel: [ val(meta), path(*.nii.gz) ]
        dti_metrics = ch_dti_metrics
            // channel: [ val(meta), path(*.nii.gz), path(*.nii.gz), ... ]
        frf = ch_fiber_response
            // channel: [ val(meta), path(*.txt) ]
        fodf = RECONST_FODF.out.fodf
            // channel: [ val(meta), path(*.nii.gz) ]
        local_tractogram = ch_local_tractogram
            // channel: [ val(meta), path(*.trk), path(*.json) ]
        pft_tractogram = ch_pft_tractogram
            // channel: [ val(meta), path(*.trk), path(*.json) ]
        versions       = ch_versions
            // channel: [ path(versions.yml) ]
}

