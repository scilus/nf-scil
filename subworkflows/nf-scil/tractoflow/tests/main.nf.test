// TODO nf-core: Once you have added the required tests, please run the following command to build this file:
// nf-core subworkflows test tractoflow
nextflow_workflow {

    name "Test Subworkflow TRACTOFLOW"
    script "../main.nf"
    workflow "TRACTOFLOW"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/tractoflow"
    // Included subworkflows
    tag "anatomical_segmentation"
    tag "preproc_dwi"
    tag "preproc_t1"
    tag "registration"
    // Included modules
    tag "reconst/dtimetrics"
    tag "registration/antsapplytransforms"
    tag "reconst/frf"
    tag "reconst/meanfrf"
    tag "reconst/dtimetrics"
    tag "reconst/fodf"
    tag "tracking/pfttracking"
    tag "tracking/localtracking"

    tag "load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from(
                    ["antsbet.zip" , "raw_T1w.zip", "raw_DWIss1000-dir32.zip", "raw_EPI.zip"]
                )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("tractoflow - REV B0") {

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        raw_t1w: it.simpleName == "raw_T1w"
                        raw_dwi: it.simpleName == "raw_DWIss1000-dir32"
                        raw_epi: it.SimpleName == "raw_EPI"
                    }
                input[0] = ch_split_test_data.raw_dwi.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]}
                input[1] = Channel.from( [] )
                input[2] = Channel.from( [] )
                input[3] = ch_split_test_data.raw_epi.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/epi.nii.gz")
                    ]}
                input[4] = ch_split_test_data.raw_t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[5] = Channel.from( [] )
                input[6] = Channel.from( [] )
                input[7] = Channel.from( [] )
                input[8] = ch_split_test_data.antsbet.map{
                    test_data_directory -> file("\${test_data_directory}/t1_template.nii.gz")}
                input[9] = ch_split_test_data.antsbet.map{
                    test_data_directory -> file("\${test_data_directory}/t1_brain_probability_map.nii.gz")}
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
                //TODO nf-core: Add all required assertions to verify the test output.
            )
        }
    }
}
