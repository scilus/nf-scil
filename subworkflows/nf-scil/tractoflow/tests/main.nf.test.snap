{
    "tractoflow - REV B0": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_dwi_resampled.nii.gz:md5,fe1633fa125292bf05fa3f72664e0956",
                        "test__bval_eddy:md5,7995daabbd74fcf7c365a39a779f67e9",
                        "test__dwi_eddy_corrected.bvec:md5,209e0c8e5800cffdce1a06b9748d133d"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_mask_resampled.nii.gz:md5,db4da2182f193c875f2d959e1b222ad1"
                    ]
                ],
                "10": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__local_tracking.trk:md5,f2620ec93f574b55b71ebd7be036ed8b",
                        "test__local_tracking_config.json:md5,ff2a194e24c4eb1c8e994be029e03c6a"
                    ]
                ],
                "11": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__pft_tracking.trk:md5,6001de533bb1335eec155f480b81a1be",
                        "test__pft_tracking_config.json:md5,1185d2d6919687f1b299eda667f6aa74"
                    ]
                ],
                "12": [
                    "versions.yml:md5,3a3c674e9221f19ed409a1890bc2eb39",
                    "versions.yml:md5,402c3604be37f571df13f983b1b5dd68",
                    "versions.yml:md5,689514aa26a4f84f83063f34c0e38190",
                    "versions.yml:md5,a3d1c964db9fa1079b738398a64219f7",
                    "versions.yml:md5,a986b4e17262e999e378666d281e9ef3",
                    "versions.yml:md5,aaab00d1fa7d070cc7ee67fd768416b8",
                    "versions.yml:md5,ad06aa4db03ace5de5712b3fa021bf89",
                    "versions.yml:md5,ba57313cd97586c885868a3aef7314a3",
                    "versions.yml:md5,d6c87a227ba82214dbafc880b2af4e8a",
                    "versions.yml:md5,fc6a2d3923b6ecd60b833ce5f09f8877"
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_warped.nii.gz:md5,d66a49cf2677670863493e0e0f9ec646"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output1Warp.nii.gz:md5,2387d68472dd2324335f6797feb64ceb",
                        "test__output0GenericAffine.mat:md5,812e708d6185c893f84326972f1fbb38",
                        "test__output0GenericAffine.mat:md5,812e708d6185c893f84326972f1fbb38",
                        "test__output1InverseWarp.nii.gz:md5,d44c7b3e88a46e89740d88b48437f839"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,b09846866983a7651b1b1bd79cb7297c",
                        "test__mask_gm.nii.gz:md5,13849fa2ed53e5e0c06dbe1f153fa7ac",
                        "test__mask_csf.nii.gz:md5,fc384dbe137d014f146b46b28078f029"
                    ]
                ],
                "5": [
                    
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__tensor.nii.gz:md5,b7bcb68dc32900e5c812ec6189a1f94f"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__ad.nii.gz:md5,e8a1b4f002daf97e45dc290c4dcf2492",
                        "test__rd.nii.gz:md5,1e3c61bc92e211d156d3fbcdc247ae00",
                        "test__md.nii.gz:md5,d72679cda1e925964bdc9be27fa3685a",
                        "test__fa.nii.gz:md5,9a06e5a37fda344e637c773e408ab720",
                        "test__ga.nii.gz:md5,ee545fe9cd38685f191381f92462527a",
                        "test__rgb.nii.gz:md5,e61698606adabce6fa95aed388165655",
                        "test__evecs_v1.nii.gz:md5,e600fca9915143c5c1dc252d20a4c3b5",
                        "test__evecs_v2.nii.gz:md5,765542d64f0d259d95d454ee3d4d9358",
                        "test__evecs_v3.nii.gz:md5,cbb35b4ac0dbe39fa585847a1630448f",
                        "test__evals_e1.nii.gz:md5,e8a1b4f002daf97e45dc290c4dcf2492",
                        "test__evals_e2.nii.gz:md5,4fbdadf345eacda17008b19d1bc94732",
                        "test__evals_e3.nii.gz:md5,906a9ad50cb25c530b8d2bef3480cac4",
                        "test__residual.nii.gz:md5,d2ae615dd84477f63485a8d89e679d7f",
                        "test__nonphysical.nii.gz:md5,b9d44001ac4c983344e1e55e60fdbac8",
                        "test__mode.nii.gz:md5,b5d07ecef283a8640413d389f0f790f1",
                        "test__norm.nii.gz:md5,f5ffd0cfaca75b2d6d1b4d5416ccb1e4"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__frf.txt:md5,94357f0520544a6c7b82a3c9605a80b3",
                        [
                            
                        ],
                        [
                            
                        ]
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fodf.nii.gz:md5,405ee57627aa837c16e3db46daacea43"
                    ]
                ],
                "brain_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_mask_resampled.nii.gz:md5,db4da2182f193c875f2d959e1b222ad1"
                    ]
                ],
                "dti_metrics": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__ad.nii.gz:md5,e8a1b4f002daf97e45dc290c4dcf2492",
                        "test__rd.nii.gz:md5,1e3c61bc92e211d156d3fbcdc247ae00",
                        "test__md.nii.gz:md5,d72679cda1e925964bdc9be27fa3685a",
                        "test__fa.nii.gz:md5,9a06e5a37fda344e637c773e408ab720",
                        "test__ga.nii.gz:md5,ee545fe9cd38685f191381f92462527a",
                        "test__rgb.nii.gz:md5,e61698606adabce6fa95aed388165655",
                        "test__evecs_v1.nii.gz:md5,e600fca9915143c5c1dc252d20a4c3b5",
                        "test__evecs_v2.nii.gz:md5,765542d64f0d259d95d454ee3d4d9358",
                        "test__evecs_v3.nii.gz:md5,cbb35b4ac0dbe39fa585847a1630448f",
                        "test__evals_e1.nii.gz:md5,e8a1b4f002daf97e45dc290c4dcf2492",
                        "test__evals_e2.nii.gz:md5,4fbdadf345eacda17008b19d1bc94732",
                        "test__evals_e3.nii.gz:md5,906a9ad50cb25c530b8d2bef3480cac4",
                        "test__residual.nii.gz:md5,d2ae615dd84477f63485a8d89e679d7f",
                        "test__nonphysical.nii.gz:md5,b9d44001ac4c983344e1e55e60fdbac8",
                        "test__mode.nii.gz:md5,b5d07ecef283a8640413d389f0f790f1",
                        "test__norm.nii.gz:md5,f5ffd0cfaca75b2d6d1b4d5416ccb1e4"
                    ]
                ],
                "dti_tensor": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__tensor.nii.gz:md5,b7bcb68dc32900e5c812ec6189a1f94f"
                    ]
                ],
                "dwi": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_dwi_resampled.nii.gz:md5,fe1633fa125292bf05fa3f72664e0956",
                        "test__bval_eddy:md5,7995daabbd74fcf7c365a39a779f67e9",
                        "test__dwi_eddy_corrected.bvec:md5,209e0c8e5800cffdce1a06b9748d133d"
                    ]
                ],
                "fodf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fodf.nii.gz:md5,405ee57627aa837c16e3db46daacea43"
                    ]
                ],
                "frf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__frf.txt:md5,94357f0520544a6c7b82a3c9605a80b3",
                        [
                            
                        ],
                        [
                            
                        ]
                    ]
                ],
                "local_tractogram": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__local_tracking.trk:md5,f2620ec93f574b55b71ebd7be036ed8b",
                        "test__local_tracking_config.json:md5,ff2a194e24c4eb1c8e994be029e03c6a"
                    ]
                ],
                "pft_tractogram": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__pft_tracking.trk:md5,6001de533bb1335eec155f480b81a1be",
                        "test__pft_tracking_config.json:md5,1185d2d6919687f1b299eda667f6aa74"
                    ]
                ],
                "t1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_warped.nii.gz:md5,d66a49cf2677670863493e0e0f9ec646"
                    ]
                ],
                "t1_to_b0_transformations": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output1Warp.nii.gz:md5,2387d68472dd2324335f6797feb64ceb",
                        "test__output0GenericAffine.mat:md5,812e708d6185c893f84326972f1fbb38",
                        "test__output0GenericAffine.mat:md5,812e708d6185c893f84326972f1fbb38",
                        "test__output1InverseWarp.nii.gz:md5,d44c7b3e88a46e89740d88b48437f839"
                    ]
                ],
                "tissues_maps": [
                    
                ],
                "tissues_masks": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,b09846866983a7651b1b1bd79cb7297c",
                        "test__mask_gm.nii.gz:md5,13849fa2ed53e5e0c06dbe1f153fa7ac",
                        "test__mask_csf.nii.gz:md5,fc384dbe137d014f146b46b28078f029"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3a3c674e9221f19ed409a1890bc2eb39",
                    "versions.yml:md5,402c3604be37f571df13f983b1b5dd68",
                    "versions.yml:md5,689514aa26a4f84f83063f34c0e38190",
                    "versions.yml:md5,a3d1c964db9fa1079b738398a64219f7",
                    "versions.yml:md5,a986b4e17262e999e378666d281e9ef3",
                    "versions.yml:md5,aaab00d1fa7d070cc7ee67fd768416b8",
                    "versions.yml:md5,ad06aa4db03ace5de5712b3fa021bf89",
                    "versions.yml:md5,ba57313cd97586c885868a3aef7314a3",
                    "versions.yml:md5,d6c87a227ba82214dbafc880b2af4e8a",
                    "versions.yml:md5,fc6a2d3923b6ecd60b833ce5f09f8877"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-29T04:11:59.683558"
    }
}